python_modules/dagster/dagster/_grpc/server.py

```diff
@@ -113,11 +117,49 @@
 
 STREAMING_CHUNK_SIZE = 4000000
 
+_UTILIZATION_METRICS = defaultdict(dict)
+_METRICS_LOCK = threading.Lock()
+METRICS_RETRIEVAL_FUNCTIONS = set()
+
+
+def _gen_request_id():
+    return str(ulid.new().uuid)
+
 
 class CouldNotBindGrpcServerToAddress(Exception):
     pass
 
 
+def retrieve_metrics(arg_from_request: str):
+    class _MetricsRetriever:
+        def __call__(self, fn: Callable) -> Callable:
+            api_call = fn.__name__
+            METRICS_RETRIEVAL_FUNCTIONS.add(api_call)
+
+            def wrapper(self, request, context):
+                _METRICS_LOCK.acquire()
+                try:
+                    req_id = _gen_request_id()
+                    _UTILIZATION_METRICS[api_call][req_id] = getattr(request, arg_from_request)
+                finally:
+                    _METRICS_LOCK.release()
+
+                res = fn(self, request, context)
+
+                _METRICS_LOCK.acquire()
+                try:
+                    del _UTILIZATION_METRICS[arg_from_request][req_id]
+                finally:
+                    _METRICS_LOCK.release()
```

Comment:
sensors and schedules are called in parallel pretty regularly and I'm a little bit worried about adding a shared lock here affecting performance. Do you think that's a reasonable concern / is there anything we could do to derisk that? It's admittedly not doing very much work inside the lock

Reply:
it would be good to have a way to opt in or out of this - so if it does cause contention issues we would have an escape hatch

Reply:
the issue is that if we want to record info about the request, then we need to record during the request execution. I think using an in mem dictionary like we do here would be faster than I/O, but if we want to avoid lock-based contention, we could use some sort of file system management; ie a directory with unique files for each request execution. Using a UUID should allow us to avoid requiring a lock.

Worth noting that the size of the dictionary is limited by the size of threadpool. I think in order for the contention to take a non negligible amount of time, you would need to reach critical mass on the dictionary.

Reply:
Having an escape hatch definitely sounds like a good idea to me.